### ISSUE TITLE
Critical 502 Bad Gateway

### ISSUE THREAD

**Issue by Tester98** [mholt/caddy#1345]
(Are you asking for help with using Caddy? Please use our forum instead: https://forum.caddyserver.com. If you are filing a bug report, please answer the following questions. If your issue is not a bug report, you do not need to use this template. Either way, please consider donating if we've helped you. Thanks!)

### 1. What version of Caddy are you running (`caddy -version`)?
Caddy 0.9.4

### 2. What are you trying to do?
Simple reverse proxy, no load balancing to another server with http2 enabled

### 3. What is your entire Caddyfile?
```text
proxy / loadbalance1:89 { 
                policy ip_hash
                fail_timeout 10s
                max_fails 1
                try_interval 250ms
                health_check /health_check
                health_check_interval 1s
                health_check_timeout 2s


                }

```

### 4. How did you run Caddy (give the full command and describe the execution environment)?
./caddy  --agree --email xyz@xyz.com

### 5. What did you expect to see?
Proxy should be working without errors  until the backend goes down

### 6. What did you see instead (give full error messages and/or log)?
502 Bad Gateway, Roughly 1-10% of the requests, Backend is not loaded at all
11/Jan/2017:01:02:45 +0700 [ERROR 502 /favicon.ico] no hosts available upstream

### 7. How can someone who is starting from scratch reproduce this behavior as minimally as possible?
Put up a simple reverse proxy and backend and maybe visit the urls 100s of times, you are bound to get one. I can confirm the behavior with firefox browser if that matters somehow.
A immediate retry gives the page successfully. Since caddy is being used in production, much logging or testing might not be possible. However, this is a serious BLOCKER. i  have located this bug  reported earlier, https://github.com/mholt/caddy/issues/1080
It has been marked as fixed. i am trying the temporary fix to disable http2 ( --http2=false) and i do get less errors. So the bug probably never got fixed. 

Any suggestions

**Comment by tw4452852** [comment]
Then this issue only happens when enable http2?

**Comment by Tester98** [comment]
We are waiting more time to confirm  whether the temporary fix gives errors or not ( -http2=false ).However,  few hours have passed and we are yet to see an error. confirmation will need 12 hrs, so we can confirm it.  we are studying the logs, sometimes we also get below lines in the errors logs.
10/Jan/2017:20:46:48 +0700 [ERROR 502 /favicon.ico] net/http: request canceled

**Comment by Tester98** [comment]
Can this be related to the https://github.com/golang/go/issues/16788, is there a way to tune so that the caddy proxy requests only http 1.1 or 1.0 requests to backends ( Disable caddy http2.0 requests to backend). However, still serving 2.0 request to end users. These bugs seem to affect almost everyone flynn, traefik and others.

**Comment by mholt** [comment]
"No hosts available upstream" means that according to your health check rules, all your backends were marked as down.  In other words, it is apparently failing health checks some of the time, or requests to it are failing. Remember, with your configuration, it only takes one failed request to mark it as down.

**Comment by Tester98** [comment]
However, After enabling the "--http2=false" , not a single 502 error has occurred, so there might be some edge case for http2 clients or backends. We are still parsing logs on both caddy and backend . backend has never reported an failed request. So our assumption is under some conditions, request never reaches the backends and still caddy returns a 502 to the client.

**Comment by Tester98** [comment]
I can confirm the bug to be http2 related, as not a single 502 has occurred, since the last 10 hrs.The only difference is that caddy is started with "--http2=false". We will be disabling http2 till the bugs gets solved. However, we will still want to be http2 enabled.  Hope , this can lead to  identify the root cause , as caddy is extremely performative in regards to certain types of workloads.

**Comment by mholt** [comment]
@Tester98 That's good news, but interesting given that https://github.com/mholt/caddy/commit/6352c9054ab7a72460f6b75c0be797eaff9d0a67 landed AFTER 0.9.4 and will be in the next release. 🤔  (unless Caddy is your backend too, then I guess it's not so surprising.)

Can you please post your _entire_ unedited Caddyfile so that we can reproduce the behavior?

**Comment by Tester98** [comment]
Caddy is not the backend,  The caddy file is 10 lines  has a domain configured which is proxied back to another server, though lets encrypt is working ( thanks for that) perfectly. We have tested the configuration in without tls and found that the bug occurred roughly 1 out of 400 times we tested it with firefox and https disabled. I can also confirm the bug to affect plaintext http requests also ( non https). so replication should be easy. 

https://github.com/mholt/caddy/commit/6352c9054ab7a72460f6b75c0be797eaff9d0a67 also confirms initial assumption that the bug is before the proxy is triggered because  backend  nevers gets hits for such request and caddy decides automatically that gateway is down. there are lots of http2 related issues in the golang repo  which may also affect caddy.

**Comment by mholt** [comment]
Thanks for responding, but if there is a bug, you're welcome to fix it yourself unless you can share your Caddyfile with us. We need to see the situation you're seeing so we can pinpoint what is causing it.

**Comment by dsnet** [comment]
It's possible that #1352 is the source of this.

**Comment by Tester98** [comment]
After 5 days of testing, We can confirm that bad gateway occurs only when http2 is enabled for caddy. No, we are not using web sockets(so it cannot be https://github.com/mholt/caddy/issues/1352). So We would recommend that if someone faces 502 bad gateway for a proxy, he can disable it unless the bug analysis is  complete. Here is our setup to replicate the bug( but it is random).

1.  Caddy is setup to serve  as reverse proxy  to a http 1.1 backend
2.  Our Backend's average reply time is 50-100ms
3.  Use firefox browser to make requests. our version is 45.0.1.
4. The page has roughly 20 different assets (css,js,img)

Random browsing over a period of time gives you 502  bad Gateway and a log line similar to `15/Jan/2017:23:13:43 +0700 [ERROR 502 /favicon.ico] net/http: request canceled`

**Comment by mholt** [comment]
@Tester98 Please share your Caddyfile with us. These instructions are too vague without your exact, unredacted, unclipped Caddyfile in full.

**Comment by tw4452852** [comment]
Hi @Tester98 , could you please give my patch a try?

**Comment by mholt** [comment]
@Tester98 Any results with the patch Tw submitted?

**Comment by Tester98** [comment]
We did some testing for a brief period with the patch and didnot receive a single 502.
i was also allowed some time to test with live production traffic (since in-house testing passed successfully) . however the bug did not appear.
 So, the bug  can be assumed as fixed. However, i have been told not to deploy it, until the next caddy version is released. We can reopen this if it appears again.

**Comment by mholt** [comment]
It will go out today. Thank you!

**Comment by fey0** [comment]
Did this fix get pushed to the main site already?
Because I am seeing similar issues with a simple reverse proxy. I updated about 2 hours ago to see if I just had an old version, but I am still seeing this issue.

Backend is a duplicati webserver and if http2 is enabled connection is constantly getting lost and then reestablished after 5-30 seconds.
I am not getting anything in the error log. 

xxxx.test {
proxy / 127.0.0.1:8200

multipass {
    handles 
    mail_from webmaster
    expires 30m
    smtp_addr 
    smtp_user 
    smtp_pass 
}

errors {
    log /var/log/caddy/dupli-error.log {
        size 5 # Rotate after 50 MB
        age  30 # Keep rotated files for 30 days
        keep 5  # Keep at most 5 log files
    }
}

}`

**Comment by tobya** [comment]
with 0.9.5 it may be worth your while extending or removing the default timeouts

https://caddyserver.com/docs/timeouts

**Comment by fey0** [comment]
Havent had a timeout in 30 minutes. (it would appear after 1 min max before) Does the timeout setting only apply to http2 connections or something? Strange that turning http2 off would also fix it.

**Comment by mholt** [comment]
Could be a bug in Go 1.7 that will be fixed in Go 1.8, too.
